### PR TITLE
feat(generator): selectively generate async methods for certain RPCs

### LINKED ISSUE #7531: Selectively generate async versions of simple RPCs
The motivation for this is that we want to generate the internals of `bigtable::AdminClient`, but it relies on making an [async stub call](https://github.com/googleapis/google-cloud-cpp/blob/1cd25c596d12d69ee5ab32d955aa26fd138210d3/google/cloud/bigtable/table_admin.cc#L574).

We do not want to generate all simple async calls, as that would lead to unnecessary clutter. By simple, I mean non-streaming, non-paginated, not a long running operation.

The simplest solution is probably to add a config like [`omitted_rpcs`](https://github.com/googleapis/google-cloud-cpp/blob/1cd25c596d12d69ee5ab32d955aa26fd138210d3/generator/generator_config.proto#L23) called `async_rpcs`. Then maintain a list of `async_methods_` similar to [what we do](https://github.com/googleapis/google-cloud-cpp/blob/1cd25c596d12d69ee5ab32d955aa26fd138210d3/generator/internal/service_code_generator.cc#L72-L75) for `methods_`. Then loop over `async_methods_` in `GenerateCc`'s as needed.

(I can probably think of a less confusing name than `async_rpcs`, but none are coming to me at the moment)

### PR DESCRIPTION
**Pull request by dbolduc**
Fixes #7531 

If it helps the reviewer, I grouped individual commits by `class`. (so most commits have a generator, a golden header/source, and a golden unit test).

I made no attempt to place the `AsyncFoo` methods next to their `Foo` counterparts.

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/7540)

### CODE REVIEW

File: generator/internal/stub_generator.cc [hunk 1]
```diff
@@ -301,6 +341,36 @@ Default$stub_class_name$::Async$method_name$(
...
+        return grpc_stub_->Async$method_name$(context, request, cq);
+      },
+      request, std::move(context)))"""},
+      {IsResponseTypeEmpty, R"""(
+      .then([](future<StatusOr<google::protobuf::Empty>> f) {
+        return f.get().status();
+      }))""", ""},
+      {";\n}\n\n"}
+    },
+            // clang-format on
```
Comment (coryan) [inline generator/internal/stub_generator.cc:369]: nit: move `clang-format on` to before the raw string.

**Review by coryan — CHANGES_REQUESTED** [review]
Some nits and a question below.

**Review by dbolduc — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/7540)* status: 0 of 36 files reviewed, 4 unresolved discussions (waiting on @coryan)

___
*[generator/internal/codegen_utils.cc, line 91 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/7540#-MnK9NE9O-3Q377R55DC-r2-91:-MnNtKdd8qpxnYgthbNC:b3zlulx) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/b15a77592bc0c58f5a2a5bcff4f30e5b12d72cd5/generator/internal/codegen_utils.cc#L91)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

nit: `in` is not super descriptive, maybe `desired_args` ? Or maybe `grouped_arg` for `in` and `group_name` for `out`?
</blockquote></details>

`single_arg` -> `grouped_arg` ?
___
*[generator/internal/codegen_utils_test.cc, line 264 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/7540#-MnK9NE9O-3Q377R55DD-r2-264:-MnNnoxy6WqPbnPhyank:bxsvnzl) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/b15a77592bc0c58f5a2a5bcff4f30e5b12d72cd5/generator/internal/codegen_utils_test.cc#L264)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Maybe a test with both `omit` and `async`?  Which got me thinking, what should be the semantics of `omit_rpc=Foo,gen_async_rpc=Foo` ?
</blockquote></details>

I did not consider what would happen if `Foo` was included in both. I think what gives us the most control is to generate the async call for `Foo` and omit the synchronous one. I don't know why we would only want to offer an async version of a call... but we could if we wanted to. (This happens to be what the code does now. I added a note in `generator_config.proto` to make this more clear).

Test added.
___
*[generator/internal/service_code_generator.cc, line 66 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/7540#-MnK9NE9O-3Q377R55DI-r2-66:-MnO0IYv4zy_P-dWMZ-P:b6ca4hm) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/b15a77592bc0c58f5a2a5bcff4f30e5b12d72cd5/generator/internal/service_code_generator.cc#L66)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

nit: feels like this should be:

```cc
auto split_arg = [this](std::string const& arg) -> std::set<std::string> {
  auto l = service_vars_.find(arg);
  if (l == service_vars_.end()) return {};
  return absl::StrSplit(l->second, ',');
};
  auto const omitted_rpcs = split_arg("omitted_rpcs");
  auto const gen_async_rpcs = split_arg("gen_async_rpcs");
```

But it might need some tuning because I am not sure how `absl::StrSplit` works in this case.

</blockquote></details>

Yes, your suggestion is much cleaner.
___
*[generator/internal/service_code_generator.cc, line 101 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/7540#-MnK9NE9O-3Q377R55DI-r2-101:-MnO0M1b2ZaLgjApoJkf:byg433c) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/b15a77592bc0c58f5a2a5bcff4f30e5b12d72cd5/generator/internal/service_code_generator.cc#L101)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Hmmm... now it looks like that lambda should be a member function.
</blockquote></details>

Refactored.
___
*[generator/internal/stub_generator.cc, line 118 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/7540#-MnK9NE9O-3Q377R55DK-r2-118:-MnNn15G7dfraMHcn6M7:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/b15a77592bc0c58f5a2a5bcff4f30e5b12d72cd5/generator/internal/stub_generator.cc#L118)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

nit: consider using a raw string (`R"""(`) here.
</blockquote></details>

Done.
___
*[generator/internal/stub_generator.cc, line 369 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/7540#-MnK9NE9O-3Q377R55DK-r2-369:-MnNnfUJ1u3212Jc8HfV:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/b15a77592bc0c58f5a2a5bcff4f30e5b12d72cd5/generator/internal/stub_generator.cc#L369)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

nit: move `clang-format on` to before the raw string.
</blockquote></details>

Done.

**Review by coryan — APPROVED** [review]
:lgtm:

> Reviewed 28 of 36 files at r1, 2 of 2 files at r2, 6 of 6 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/7540#-:-MnO97-k0vuBvP5w-y9J:bnfp4nl)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @dbolduc)

___
*[generator/internal/codegen_utils.cc, line 91 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/7540#-MnK9NE9O-3Q377R55DC-r2-91:-MnO8OozA-C2fJhq_NJv:b-80tlqk) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/b15a77592bc0c58f5a2a5bcff4f30e5b12d72cd5/generator/internal/codegen_utils.cc#L91)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

`single_arg` -> `grouped_arg` ?
</blockquote></details>

Sure.

**Pull request merged**